lib/acts_as_xapian/acts_as_xapian.rb

```diff
@@ -611,6 +611,69 @@ def log_description
   class ActsAsXapianJob < ActiveRecord::Base
   end
 
+  class FailedJob
+    attr_reader :job_id, :error, :model_data
+
+    def initialize(job_id, error, model_data = {})
+      @job_id = job_id
+      @error = error
+      @model_data = model_data
+    end
+
+    def job_info
+      msg = "FAILED ActsAsXapian.update_index job #{ job_id } #{ error.class }"
+      msg += " model #{ job_model }" if job_model
+      msg += " id #{ job_model_id }" if job_model_id
+      msg += '.'
+      msg
+    end
+
+    def error_backtrace
+      error.backtrace.join("\n")
```

Comment:
For some strange reason I'm not getting backtraces for exceptions thrown on line [749](https://github.com/mysociety/alaveteli/blob/c5f54072f7fdbdeed2efed22beed23d17824fa30/lib/acts_as_xapian/acts_as_xapian.rb#L749) so this line blows up which is reeeally hard to track down, the full error and stack given was:

```
NoMethodError: undefined method `join' for nil:NilClass
Traceback (most recent call last):
./bin/rake: undefined method `reject' for nil:NilClass (NoMethodError)
```

Probably best to change this line to something like:

`error.backtrace.try(:join, "\n").to_s`

Reply:
Hmm strange, I get:

```
$ script/update-xapian-index
FAILED ActsAsXapian.update_index job 643 RuntimeError model InfoRequestEvent id 900.

This job will be removed from the queue. Once the underlying problem is fixed, manually re-index the model record.

You can do this in a rails console with `InfoRequestEvent.find(900).xapian_mark_needs_index`.

---------
Backtrace
---------

/home/vagrant/alaveteli/lib/acts_as_xapian/acts_as_xapian.rb:749:in `run_job'
/home/vagrant/alaveteli/lib/acts_as_xapian/acts_as_xapian.rb:713:in `block (2 levels) in update_index'
/home/vagrant/bundle/ruby/2.3.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
/home/vagrant/bundle/ruby/2.3.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/transaction.rb:184:in `within_new_transaction'
/home/vagrant/bundle/ruby/2.3.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
/home/vagrant/bundle/ruby/2.3.0/gems/activerecord-4.2.10/lib/active_record/transactions.rb:220:in `transaction'
/home/vagrant/alaveteli/lib/acts_as_xapian/acts_as_xapian.rb:702:in `block in update_index'
/home/vagrant/alaveteli/lib/acts_as_xapian/acts_as_xapian.rb:699:in `each'
/home/vagrant/alaveteli/lib/acts_as_xapian/acts_as_xapian.rb:699:in `update_index'
/home/vagrant/alaveteli/lib/acts_as_xapian/tasks/xapian.rake:12:in `block (2 levels) in <top (required)>'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/task.rb:251:in `block in execute'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/task.rb:251:in `each'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/task.rb:251:in `execute'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/task.rb:195:in `block in invoke_with_call_chain'
/usr/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/task.rb:188:in `invoke_with_call_chain'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/task.rb:181:in `invoke'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/application.rb:153:in `invoke_task'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/application.rb:109:in `block (2 levels) in top_level'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/application.rb:109:in `each'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/application.rb:109:in `block in top_level'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/application.rb:118:in `run_with_threads'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/application.rb:103:in `top_level'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/application.rb:81:in `block in run'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/application.rb:179:in `standard_exception_handling'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/lib/rake/application.rb:78:in `run'
/home/vagrant/bundle/ruby/2.3.0/gems/rake-12.1.0/exe/rake:27:in `<top (required)>'
/home/vagrant/bundle/ruby/2.3.0/bin/rake:22:in `load'
/home/vagrant/bundle/ruby/2.3.0/bin/rake:22:in `<main>'
```

To get that, I did:

```sh
bin/rails runner "ActsAsXapian::ActsAsXapianJob.create(model: 'InfoRequestEvent', model_id: InfoRequestEvent.last.id, action: 'wjheebwc')"
script/update-xapian-index
```

Reply:
Difference seems to be cause by Ruby versions; 2.3.x I get the same as you, 2.5.x I get what I did yesterday.

Only relevant change in 2.5 I know about is exception backtraces are now printed in reverse order - doesn't seem like that would cause this problem.

Installing 2.4 now.

Reply:
2.4.x has a bracktrace

Reply:
Yeah, seems like [`backtrace`](http://ruby-doc.org/core-2.5.0/Exception.html#method-i-backtrace) is still a method on `Exception `.

Also, looks like you can [`set_backtrace`](http://ruby-doc.org/core-2.0.0/Exception.html#method-i-set_backtrace) rather than stubbing it out. TIL.

Reply:
So seems to be an issue with 2.5, with Rails 4.2:

I've got a reduced test case proving an issue:

```ruby
namespace :temp do
  task :foo do
    raise 'foo'
  end
end
```

This throws a similar error.  But if in `Rakefile`, if I remove the Rails' `require` and its `load_tasks` the task works.